### PR TITLE
Refactoring. Move getDrawingSurface to Java from JNI

### PR DESCRIPTION
**Pull request by igordmn**

### CODE REVIEW

File: skiko/src/jvmMain/cpp/common/awt_jni.cc [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+extern "C"
+{
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_AWTKt_getAWT(JNIEnv *env, jobject obj)
+    {
+        JAWT *awt = new JAWT();
+        awt->version = (jint)JAWT_VERSION_9;
+        jboolean result = Skiko_GetAWT(env, awt);
+
+        if (result == JNI_FALSE)
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/awt_jni.cc:14]: `if (result == JNI_FALSE) {` I presume.

File: skiko/src/jvmMain/cpp/common/awt_jni.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_AWTKt_getDrawingSurface(JNIEnv *env, jobject obj, jlong awtPtr, jobject layer)
+    {
+        JAWT *awt = reinterpret_cast<JAWT *>(static_cast<uintptr_t>(awtPtr));
+        JAWT_DrawingSurface *ds = awt->GetDrawingSurface(env, layer);
+        return static_cast<jlong>(reinterpret_cast<uintptr_t>(ds));
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_AWTKt_freeDrawingSurface(JNIEnv *env, jobject obj, jlong awtPtr, jlong drawingSurfacePtr)
+    {
+        JAWT *awt = reinterpret_cast<JAWT *>(static_cast<uintptr_t>(awtPtr));
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/awt_jni.cc:31]: Maybe make generic templates like 
```
template<typename T>
T* toType(jlong ptr) { return  reinterpret_cast<T*>(static_cast<uintptr_t>(ptr)); }
template<typename T>
jlong fromType(T* ptr) { return  static_cast<jlong>(reinterpret_cast<uintptr_t>(ptr)); }
```

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/AWT.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            return block()
+        } finally {
+            unlock()
+        }
+    }
+
+    fun getInfo() = DrawingSurfaceInfo(ptr)
+
+    override fun close() {
+        freeDrawingSurface(awt, ptr)
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/AWT.kt:48]: Shalln't you clear `ptr`?

File: skiko/src/jvmMain/objectiveC/macos/drawlayer.m [hunk 4]
```diff
@@ -81,64 +81,37 @@ - (void) makeFullscreen: (BOOL) value
 
 extern jboolean Skiko_GetAWT(JNIEnv* env, JAWT* awt);
 
-JNIEXPORT void JNICALL Java_org_jetbrains_skiko_HardwareLayer_init(JNIEnv *env, jobject canvas)
+JNIEXPORT void JNICALL Java_org_jetbrains_skiko_HardwareLayer_nativeInit(JNIEnv *env, jobject canvas, jlong platformInfoPtr)
 {
     if (layerStorage == nil)
     {
         layerStorage = [[NSMutableSet alloc] init];
```
Comment (olonho) [inline skiko/src/jvmMain/objectiveC/macos/drawlayer.m:88]: Couldn't we get rid of `layerStorage` global and pass it explicitly from Kotlin side?

Comment (igordmn) [reply skiko/src/jvmMain/objectiveC/macos/drawlayer.m:88]: We can, but it should be done in another PR, these changes don't fit this PR (it is only about getAWT/getDrawingSurface) and I am afraid to break this :)

File: skiko/src/jvmMain/cpp/common/awt_jni.cc (hunk 1 above)
Comment (igordmn) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc:15]: I think we need to preserve the current style (but I like Java-style more :) )

File: skiko/src/jvmMain/cpp/common/awt_jni.cc (hunk 2 above)
Comment (igordmn) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc]: Great! Implemented these templates.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/AWT.kt (hunk 3 above)
Comment (igordmn) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/AWT.kt:50]: Agree

File: skiko/src/jvmMain/cpp/common/awt_jni.cc (hunk 1 above)
Comment (olonho) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc:15]: Nevermind, we likely need to reformat the whole codebase with clang-format.

File: skiko/src/jvmMain/objectiveC/macos/drawlayer.m (hunk 4 above)
Comment (olonho) [reply skiko/src/jvmMain/objectiveC/macos/drawlayer.m:86]: OK.

File: skiko/src/jvmMain/cpp/common/awt_jni.cc [hunk 5]
```diff
@@ -17,53 +18,53 @@ extern "C"
...
+        JAWT_DrawingSurface *ds = fromJavaPointer<JAWT_DrawingSurface *>(drawingSurfacePtr);
+        JAWT_DrawingSurfaceInfo *dsi = fromJavaPointer<JAWT_DrawingSurfaceInfo *>(drawingSurfaceInfoPtr);
         ds->FreeDrawingSurfaceInfo(dsi);
     }
 
     JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_AWTKt_getPlatformInfo(JNIEnv *env, jobject obj, jlong drawingSurfaceInfoPtr)
     {
-        JAWT_DrawingSurfaceInfo *dsi = reinterpret_cast<JAWT_DrawingSurfaceInfo *>(static_cast<uintptr_t>(drawingSurfaceInfoPtr));
-        return static_cast<jlong>(reinterpret_cast<uintptr_t>(dsi->platformInfo));
+        JAWT_DrawingSurfaceInfo *dsi = fromJavaPointer<JAWT_DrawingSurfaceInfo *>(drawingSurfaceInfoPtr);
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/awt_jni.cc:67]: I'd do those without '*' here moving it to the template.

File: skiko/src/jvmMain/cpp/common/jni_helpers.h [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+template<typename T>
+T fromJavaPointer(jlong ptr) { return reinterpret_cast<T>(static_cast<uintptr_t>(ptr)); }
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/jni_helpers.h:4]: You could also make those inline.

Comment (olonho) [reply skiko/src/jvmMain/cpp/common/jni_helpers.h:4]: And return T*

File: skiko/src/jvmMain/cpp/common/awt_jni.cc (hunk 5 above)
Comment (igordmn) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc:67]: There a types (aliases of long) without * (HDC, for example)

File: skiko/src/jvmMain/cpp/common/jni_helpers.h (hunk 6 above)
Comment (igordmn) [reply skiko/src/jvmMain/cpp/common/jni_helpers.h]: > You could also make those inline.
Done

> And return T*

see https://github.com/JetBrains/skiko/pull/62#discussion_r569530245

File: skiko/src/jvmMain/cpp/common/awt_jni.cc (hunk 5 above)
Comment (olonho) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc:67]: IIRC those are typedef'ed as HANDLE which is void*.

Comment (olonho) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc:67]: so `fromJavaPointer<void>` would work for them. Anyway, matter of taste, feel free to do it the way you like.

Comment (igordmn) [reply skiko/src/jvmMain/cpp/common/awt_jni.cc:67]: But we should use "HANDLE" as a template parameter, not "void"

File: skiko/src/jvmMain/cpp/linux/drawlayer.cc [hunk 7]
```diff
@@ -7,52 +7,24 @@
 #include <cstdlib>
 #include <unistd.h>
 #include <stdio.h>
+#include "../common/jni_helpers.h"
```
[the remark this answers is not in the document]
Comment (igordmn) [reply skiko/src/jvmMain/cpp/linux/drawlayer.cc]: Done

**Pull request merged**